populate_keyspace: use datadir

bhalevy opened the pull request:
Currently the datadir is ignored.
Use it to construct the table's base path.

Fixes scylladb/scylladb#15418

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3790/

denesb reviewed (approved):
Can we also add a unit test to ensure this doesn't get broken in the future?

bhalevy commented on replica/distributed_loader.cc:
```diff
@@ -300,13 +301,19 @@ class table_populator {
-    {}
+    {
+        _base_path = database::table_dir(
+            database::keyspace_dir(_datadir, _ks),
```
ok, not extremely elegant since the caller will have to find the right one based on its prefix, but that could work.

xemul replied:
No, no, I meant something different.

Right now the population loop looks like

```
    for ks in db.keyspaces {
        for dir in cfg.data_dirs {
            for table in ks.tables {
                populate_table(path(dir / ks / table.basename()))
            }
        }
    }
```

but it could look simple, like this

```
    for ks in db.keyspaces {
        for table in ks.tables {
            for dir in table.datadirs() {
                populate_table(path(dir))
            }
        }
    }
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3848/

xemul commented on replica/distributed_loader.cc:
```diff
@@ -460,25 +460,20 @@ future<> table_populator::populate_subdir(sstables::sstable_state state, allow_o
-        dblog.info("Keyspace {}: Reading CF {} id={} version={} storage={}", ks_name, cfname, uuid, s->version(), cf->get_storage_options().type_string());
+      // might have more than one dir for a keyspace iff data_file_directories is > 1 and
+      // somehow someone placed sstables in more than one of them for a given ks. (import?)
+      co_await coroutine::parallel_for_each(db.local().get_config().data_file_directories(), [&] (const sstring& datadir) -> future<> {
```
Why don't you want to get_table_on_all_shards() first and then walk its `_config.all_datadirs` instead?

bhalevy replied:
Good idea.  Will do.

bhalevy commented:
In [7bd131d](https://github.com/scylladb/scylladb/commit/7bd131d212e1a1d7f549d91f6d9887cee1946b2a):
- get global table ptr in the outer loop
- loop on all_datadirs in the inner loop as per https://github.com/scylladb/scylladb/pull/15480#discussion_r1334245766
- cleaned up the code a bit
- added: "distributed_loader: populate_keyspace: access cf by ref"

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3886/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3889/
